File: compiler/validator.py
"""Workflow IR Validator — 校验 IR 完整性

第二阶段：校验 IRWorkflow 的结构完整性和语义正确性。
MVP1 校验：必填字段、步骤 ID 唯一性、类型字段匹配。
MVP2 扩展：条件分支字段完整性、引用步骤ID存在性、表达式语法校验。
"""

import re

from flowforge.compiler.ir import IRWorkflow, StepType


class WorkflowValidator:
    """IRWorkflow 校验器"""

    # 条件表达式基本语法：${{...}} 或 ${...} 包裹的表达式
    _CONDITION_PATTERN = re.compile(r"^\$\{.+\}$")

    def validate(self, workflow: IRWorkflow) -> list[str]:
        """校验 IR，返回错误列表（空列表表示通过）

        Args:
            workflow: 待校验的 IRWorkflow 实例。

        Returns:
            错误信息列表，空列表表示校验通过。
        """
        errors: list[str] = []

        # 必填字段校验
        if not workflow.id:
            errors.append("Workflow id is required")
        if not workflow.name:
            errors.append("Workflow name is required")
        if not workflow.steps:
            errors.append("Workflow must have at least one step")

        # 第一遍：收集所有步骤ID + 基础校验
        step_ids: set[str] = set()
        for step in workflow.steps:
            if step.id in step_ids:
                errors.append(f"Duplicate step id: {step.id}")
            step_ids.add(step.id)

            # 类型字段匹配校验
            if step.type == StepType.AGENT and not step.agent:
                errors.append(f"Step '{step.id}': agent type requires 'agent' field")
            if step.type == StepType.TOOL and not step.tool:
                errors.append(f"Step '{step.id}': tool type requires 'tool' field")

        # 第二遍：条件分支校验（需要完整的 step_ids 集合）
        for step in workflow.steps:
            errors.extend(self._validate_conditional(step, step_ids))

        # 第三遍：PARALLEL / FALLBACK / LOOP 校验
        for step in workflow.steps:
            errors.extend(self._validate_parallel(step))
            errors.extend(self._validate_fallback(step))
            errors.extend(self._validate_loop(step))

        return errors

    def _validate_conditional(self, step, all_step_ids: set[str]) -> list[str]:
        """校验条件分支字段的完整性和语义正确性"""
        errors: list[str] = []
        has_condition = step.condition is not None
        has_on_true = step.on_true is not None
        has_on_false = step.on_false is not None

        # condition 存在时，on_true 和 on_false 必须同时存在
        if has_condition:
            if not has_on_true:
                errors.append(
                    f"Step '{step.id}': condition requires 'on_true' field"
                )
            if not has_on_false:
                errors.append(
                    f"Step '{step.id}': condition requires 'on_false' field"
                )

            # 条件表达式语法校验：必须以 ${ 开头和 } 结尾
            if not self._CONDITION_PATTERN.match(step.condition):
                errors.append(
                    f"Step '{step.id}': condition expression must be wrapped in "
                    f"${{...}} syntax, got: {step.condition}"
                )

        # on_true/on_false 存在时，condition 必须存在
        if (has_on_true or has_on_false) and not has_condition:
            errors.append(
                f"Step '{step.id}': 'on_true'/'on_false' requires 'condition' field"
            )

        # on_true/on_false 引用的步骤ID必须存在
        if has_on_true and step.on_true not in all_step_ids:
            errors.append(
                f"Step '{step.id}': on_true references non-existent step '{step.on_true}'"
            )
        if has_on_false and step.on_false not in all_step_ids:
            errors.append(
                f"Step '{step.id}': on_false references non-existent step '{step.on_false}'"
            )

        # on_false 不能引用自身（无条件跳回自身是死循环）
        # 但 on_true 引用自身是合法的（表示条件满足时循环执行，如"分数不够就重写"）
        if has_on_false and step.on_false == step.id:
            errors.append(
                f"Step '{step.id}': on_false cannot reference itself (infinite loop)"
            )

        return errors

    def _validate_parallel(self, step) -> list[str]:
        """校验 PARALLEL 步骤的完整性"""
        errors: list[str] = []
        if step.type != StepType.PARALLEL:
            return errors
        if not step.parallel_steps:
            errors.append(
                f"Step '{step.id}': parallel type requires 'parallel_steps' field with at least one step"
            )
        return errors

    def _validate_fallback(self, step) -> list[str]:
        """校验 FALLBACK 步骤的完整性"""
        errors: list[str] = []
        if step.type != StepType.FALLBACK:
            return errors
        if not step.primary:
            errors.append(
                f"Step '{step.id}': fallback type requires 'primary' field with at least one step"
            )
        if not step.fallback:
            errors.append(
                f"Step '{step.id}': fallback type requires 'fallback' field with at least one step"
            )
        return errors

    def _validate_loop(self, step) -> list[str]:
        """校验 LOOP 步骤的完整性"""
        errors: list[str] = []
        if step.type != StepType.LOOP:
            return errors
        if not step.loop_steps:
            errors.append(
                f"Step '{step.id}': loop type requires 'loop_steps' field with at least one step"
            )
        if step.max_iterations < 1:
            errors.append(
                f"Step '{step.id}': loop type requires 'max_iterations' >= 1, got {step.max_iterations}"
            )
        return errors

File: packages/forgekin/workflow-compiler/src/validator.ts
/**
 * @flowforge/forgekin-workflow-compiler — 工作流 IR 校验器
 *
 * 校验 IRWorkflow 结构完整性与语义正确性（对齐 Python `compiler/validator.py`）：
 * - 必填字段 / 步骤 ID 唯一性 / 类型字段匹配（agent 需 agent、tool 需 tool）
 * - MVP2 条件分支：condition↔on_true/on_false 配套、`${...}` 语法、引用步骤存在性、
 *   on_false 不能引用自身（on_true 自引用合法：条件满足时循环执行）
 * - MVP3 PARALLEL / FALLBACK / LOOP 完整性
 */
import { IRStep, IRWorkflow } from './ir.js';

/** 条件表达式基本语法：${...} 包裹（对齐 Python _CONDITION_PATTERN） */
const CONDITION_PATTERN = /^\$\{.+\}$/;

export class WorkflowValidator {
  /** 校验 IR，返回错误列表（空列表表示通过） */
  validate(workflow: IRWorkflow): string[] {
    const errors: string[] = [];

    // 必填字段校验
    if (!workflow.id) {
      errors.push('Workflow id is required');
    }
    if (!workflow.name) {
      errors.push('Workflow name is required');
    }
    if (!workflow.steps || workflow.steps.length === 0) {
      errors.push('Workflow must have at least one step');
    }

    // 第一遍：收集所有步骤 ID + 基础校验
    const stepIds = new Set<string>();
    for (const step of workflow.steps) {
      if (stepIds.has(step.id)) {
        errors.push(`Duplicate step id: ${step.id}`);
      }
      stepIds.add(step.id);

      // 类型字段匹配校验
      if (step.type === 'agent' && !step.agent) {
        errors.push(`Step '${step.id}': agent type requires 'agent' field`);
      }
      if (step.type === 'tool' && !step.tool) {
        errors.push(`Step '${step.id}': tool type requires 'tool' field`);
      }
    }

    // 第二遍：条件分支校验（需要完整 step_ids）
    for (const step of workflow.steps) {
      errors.push(...this.validateConditional(step, stepIds));
    }

    // 第三遍：PARALLEL / FALLBACK / LOOP 校验
    for (const step of workflow.steps) {
      errors.push(...this.validateParallel(step));
      errors.push(...this.validateFallback(step));
      errors.push(...this.validateLoop(step));
    }

    return errors;
  }

  /** 校验条件分支字段的完整性和语义正确性 */
  validateConditional(step: IRStep, allStepIds: Set<string>): string[] {
    const errors: string[] = [];
    const hasCondition = step.condition !== undefined;
    const hasOnTrue = step.onTrue !== undefined;
    const hasOnFalse = step.onFalse !== undefined;

    // condition 存在时，on_true 和 on_false 必须同时存在
    if (hasCondition) {
      if (!hasOnTrue) {
        errors.push(`Step '${step.id}': condition requires 'on_true' field`);
      }
      if (!hasOnFalse) {
        errors.push(`Step '${step.id}': condition requires 'on_false' field`);
      }
      // 条件表达式语法校验：必须以 ${ 开头和 } 结尾
      if (step.condition !== undefined && !CONDITION_PATTERN.test(step.condition)) {
        errors.push(
          `Step '${step.id}': condition expression must be wrapped in ${'${...}'} syntax, got: ${step.condition}`,
        );
      }
    }

    // on_true/on_false 存在时，condition 必须存在
    if ((hasOnTrue || hasOnFalse) && !hasCondition) {
      errors.push(`Step '${step.id}': 'on_true'/'on_false' requires 'condition' field`);
    }

    // on_true/on_false 引用的步骤 ID 必须存在
    if (hasOnTrue && step.onTrue !== undefined && !allStepIds.has(step.onTrue)) {
      errors.push(`Step '${step.id}': on_true references non-existent step '${step.onTrue}'`);
    }
    if (hasOnFalse && step.onFalse !== undefined && !allStepIds.has(step.onFalse)) {
      errors.push(`Step '${step.id}': on_false references non-existent step '${step.onFalse}'`);
    }

    // on_false 不能引用自身（无条件跳回自身是死循环）；on_true 自引用合法
    if (hasOnFalse && step.onFalse === step.id) {
      errors.push(`Step '${step.id}': on_false cannot reference itself (infinite loop)`);
    }

    return errors;
  }

  /** 校验 PARALLEL 步骤的完整性 */
  validateParallel(step: IRStep): string[] {
    const errors: string[] = [];
    if (step.type !== 'parallel') {
      return errors;
    }
    if (!step.parallelSteps || step.parallelSteps.length === 0) {
      errors.push(
        `Step '${step.id}': parallel type requires 'parallel_steps' field with at least one step`,
      );
    }
    return errors;
  }

  /** 校验 FALLBACK 步骤的完整性 */
  validateFallback(step: IRStep): string[] {
    const errors: string[] = [];
    if (step.type !== 'fallback') {
      return errors;
    }
    if (!step.primary || step.primary.length === 0) {
      errors.push(
        `Step '${step.id}': fallback type requires 'primary' field with at least one step`,
      );
    }
    if (!step.fallback || step.fallback.length === 0) {
      errors.push(
        `Step '${step.id}': fallback type requires 'fallback' field with at least one step`,
      );
    }
    return errors;
  }

  /** 校验 LOOP 步骤的完整性 */
  validateLoop(step: IRStep): string[] {
    const errors: string[] = [];
    if (step.type !== 'loop') {
      return errors;
    }
    if (!step.loopSteps || step.loopSteps.length === 0) {
      errors.push(
        `Step '${step.id}': loop type requires 'loop_steps' field with at least one step`,
      );
    }
    if (step.maxIterations < 1) {
      errors.push(
        `Step '${step.id}': loop type requires 'max_iterations' >= 1, got ${step.maxIterations}`,
      );
    }
    return errors;
  }
}
